p2p/socket.go

```diff
@@ -0,0 +1,461 @@
+				if err != nil {
+					s.error(err)
+					break reader
+				}
+
+				if nonce != 0 {
+					_, ch, found := s.requests.Find(nonce)
+					if !found {
+						// report that we've received a nonced message whose requested does not exist on our end!
+					}
+
+					ch <- types.NewWork(nonce, data, s.addr, wrapped)
+					close(ch)
+					continue
+				}
+
+				s.runner.Sink() <- types.NewWork(nonce, data, s.addr, wrapped)
+			}
+		}
+	}
+}
+
+// TODO(derrandz): add buffered write by writing exactly the writeBufferSize. Will need to think about how to split big payloads and sequence them
+// writes a chunk=writeBufferSize to the writer
+// this operation is blocking, and blocks until a waiter is ready to receive the signal from the write buffer (signaling that is has written)
+// This is used by send/request/broadcast operations
+// upon each send, the write routine will receive a signal so that it may proceed to send the write over the network
+func (s *socket) writeChunk(b []byte, iserroreof bool, reqnum uint32, wrapped bool) (uint, error) {
+	defer s.buffers.write.Unlock()
+	s.buffers.write.Lock()
+
+	writeBuffer := s.buffers.write.Ref()
+
+	buff := s.c.encode(Binary, iserroreof, reqnum, b, wrapped)
+	(*writeBuffer) = append((*writeBuffer), buff...)
+
+	s.buffers.write.Signal()
+	return uint(len(b)), nil // TODO(derrandz): should length be of b or of the encoded b
+}
+
+// writeChunkAckful is a writeChunk that expects to receive an ACK response for the chunk it has written
+// This method will create a request, which is basically a nonce to identify the chunk to write, and a channel on which to receive the response
+// the channel is blocking, thus allowing the 'wait to receive the response' behavior.
+// the `read` routine takes care of identifying incoming responses (_using the nonce_) and redirecting them to the waiting channels of the currently-open requests.
+func (s *socket) writeChunkAckful(b []byte, wrapped bool) (types.Work, error) {
+	request := s.requests.Get()
+	requestNonce := request.Nonce()
+
+	if _, err := s.writeChunk(b, false, requestNonce, wrapped); err != nil {
+		s.requests.Delete(requestNonce)
+		return types.NewWork(requestNonce, nil, "", false), err
+	}
+
+	var response types.Work
+
+	select {
+	case response = <-request.Response():
+		return response, nil
+
+	case <-time.After(time.Millisecond * time.Duration(s.readTimeout)):
+		return types.Work{}, ErrSocketRequestTimedOut(s.addr, requestNonce)
+	}
+}
+
+// the write routine
+// this routine performs buffered writes (writeBufferSize) on the established connection
+// this routine halfs if:
+//   - the cancelable routine cancels
+//   - if the runner stops
+//   - if the socket closes
+//   - if some party signals that this routine should stop (by closing the s.writing channel)
+//   - if there is a write error
+func (s *socket) write(ctx context.Context) {
+	defer func() {
+		close(s.writing)
+		s.closed <- 1
+	}()
+
+	{
```

Comment:
Remove extra scope?

Reply:
As answered above. I really think this is comfortable for my eyes. Wdyt?